File: Python/src/cpp/modify_priors.cpp
#include <multiRL/modify_priors.hpp>

#include <stdexcept>
#include <unordered_set>

namespace multiRL {

namespace {

PriorType prior_type_from_string(const std::string& type) {
    if (type == "normal" || type == "norm") {
        return PriorType::NORMAL;
    }
    if (type == "uniform" || type == "unif") {
        return PriorType::UNIFORM;
    }
    if (type == "lognormal" || type == "lnorm") {
        return PriorType::LOGNORMAL;
    }
    if (type == "cauchy") {
        return PriorType::CAUCHY;
    }
    if (type == "beta") {
        return PriorType::BETA;
    }
    if (type == "exponential" || type == "exp") {
        return PriorType::EXPONENTIAL;
    }
    if (type == "none") {
        return PriorType::NONE;
    }

    throw std::invalid_argument("Unknown prior type.");
}

}  // namespace

/* ========================================================================== *
 *                         Prior Standardization                              *
 * ========================================================================== */

PriorGroup modify_priors(
    const std::vector<std::string>& names,
    const std::vector<std::string>& types,
    const std::vector<double>& param1,
    const std::vector<double>& param2,
    const std::vector<std::string>& free_names,
    bool active
) {
    PriorGroup out;
    out.active = active;

    if (!active) {
        return out;
    }

    if (names.size() != types.size() ||
        names.size() != param1.size() ||
        names.size() != param2.size()) {
        throw std::invalid_argument("Prior specification sizes do not match.");
    }

    std::unordered_set<std::string> free_set;
    for (const std::string& name : free_names) {
        free_set.insert(name);
    }

    for (std::size_t index = 0; index < names.size(); ++index) {
        if (free_set.find(names[index]) == free_set.end()) {
            throw std::invalid_argument(
                "Prior names must match free parameter names."
            );
        }

        PriorSpec spec;
        spec.type = prior_type_from_string(types[index]);
        spec.param1 = param1[index];
        spec.param2 = param2[index];
        out.specs[names[index]] = spec;
    }

    return out;
}

}  // namespace multiRL

File: R/R/modify_priors.R
.modify_priors <- function(priors, params) {
  if (base::length(priors) == 0L) {
    return(.modify_priors_empty())
  }

  if (!base::identical(base::names(priors), base::names(params))) {
    base::stop(
      "The names of 'priors' must be identical to the names of 'params'."
    )
  }

  specs <- base::lapply(priors, .modify_priors_to_spec)

  list(
    active = TRUE,
    name = base::names(specs),
    type = base::vapply(specs, `[[`, character(1L), "type"),
    param1 = base::vapply(specs, `[[`, numeric(1L), "param1"),
    param2 = base::vapply(specs, `[[`, numeric(1L), "param2")
  )
}

.modify_priors_empty <- function() {
  list(
    active = FALSE,
    name = character(),
    type = character(),
    param1 = numeric(),
    param2 = numeric()
  )
}

.modify_priors_to_spec <- function(prior) {
  if (base::is.list(prior) && !base::is.null(prior$type)) {
    return(.modify_priors_list_to_spec(prior))
  }

  base::stop("Priors must be structured prior lists.")
}

.modify_priors_list_to_spec <- function(prior) {
  type <- base::tolower(base::as.character(prior$type[[1L]]))

  list(
    type = .modify_priors_normalize_type(type),
    param1 = .modify_priors_arg1(prior, type),
    param2 = .modify_priors_arg2(prior, type)
  )
}

.modify_priors_normalize_type <- function(type) {
  switch(
    type,
    norm = "normal",
    normal = "normal",
    unif = "uniform",
    uniform = "uniform",
    lnorm = "lognormal",
    lognormal = "lognormal",
    cauchy = "cauchy",
    beta = "beta",
    exp = "exponential",
    exponential = "exponential",
    none = "none",
    base::stop("Unsupported prior type.")
  )
}

.modify_priors_arg1 <- function(prior, type) {
  switch(
    .modify_priors_normalize_type(type),
    normal = .modify_priors_named_number(prior, c("mean", "mu", "param1")),
    uniform = .modify_priors_named_number(prior, c("min", "lower", "param1")),
    lognormal = .modify_priors_named_number(
      prior,
      c("meanlog", "mean", "param1")
    ),
    cauchy = .modify_priors_named_number(
      prior,
      c("location", "mean", "param1")
    ),
    beta = .modify_priors_named_number(prior, c("shape1", "alpha", "param1")),
    exponential = .modify_priors_named_number(
      prior,
      c("rate", "lambda", "param1")
    ),
    none = NaN
  )
}

.modify_priors_arg2 <- function(prior, type) {
  switch(
    .modify_priors_normalize_type(type),
    normal = .modify_priors_named_number(prior, c("sd", "sigma", "param2")),
    uniform = .modify_priors_named_number(prior, c("max", "upper", "param2")),
    lognormal = .modify_priors_named_number(
      prior,
      c("sdlog", "sd", "param2")
    ),
    cauchy = .modify_priors_named_number(prior, c("scale", "sd", "param2")),
    beta = .modify_priors_named_number(prior, c("shape2", "beta", "param2")),
    exponential = NaN,
    none = NaN
  )
}

.modify_priors_named_number <- function(x, keys) {
  for (key in keys) {
    if (!base::is.null(x[[key]])) {
      return(base::as.numeric(x[[key]][[1L]]))
    }
  }

  base::stop("Structured prior list is missing a required argument.")
}
